File: api/src/main/java/club/yunzhi/workhome/service/WorkService.java
package club.yunzhi.workhome.service;

import club.yunzhi.workhome.entity.Attachment;
import club.yunzhi.workhome.entity.Work;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import javax.validation.constraints.NotNull;
import java.util.IllegalFormatException;
import java.util.List;
import java.util.Optional;

/**
 * 作业
 *
 * @author yz
 */
public interface WorkService {
    /**
     * 获取当前登录学生的作业列表
     *
     * @return 作业
     */
    List<Work> getAllOfCurrentStudent();

    /**
     * 获取当前登录学生某个实验的作业
     *
     * @param itemId 实验
     * @return 作业 | null
     */
    Optional<Work> getByItemIdOfCurrentStudent(Long itemId);

    /**
     * 获取作业
     *
     * @param itemId    实验
     * @param studentId 学生
     * @return 作业 | null
     */
    Optional<Work> getByItemIdAndStudentId(Long itemId, Long studentId);

    /**
     * 获取作业，如果获取不到就待久化一个作业
     *
     * @param itemId 学期
     * @return 作业
     */
    Work getOrElseCreateNewByItemIdOfCurrentStudent(Long itemId);

    /**
     * 更新当前学生
     *
     * @param id   作业ID
     * @param work 作业
     * @return 作业
     */
    Work updateOfCurrentStudent(Long id, Work work);

    /**
     * 新建作业
     *
     * @param itemId 学期
     * @return 作业
     */
    Work saveWorkByItemIdOfCurrentStudent(Long itemId);

    /**
     * 新建作业
     *
     * @param work 作业
     * @return 作业
     */
    Work save(Work work);

    /**
     * 删除某个作业的某个附件
     *
     * @param workId       作业id
     * @param attachmentId 附件id
     */
    void deleteAttachment(Long workId, Long attachmentId);

    /**
     * 通过id获取作业
     *
     * @param id 作业id
     * @return 作业
     */
    Work findById(Long id);

    /**
     * 上传作业并将作业保存在自己的目录下
     *
     * @param multipartFile 文件
     * @return 附件实体
     */
    Attachment uploadWork(MultipartFile multipartFile, String itemId, String uploadDir) throws IllegalFormatException;

    /**
     * 获取所有作业
     *
     * @return 所有作业
     */
    Page<Work> getAll(Pageable pageable);


    /**
     * 更新作业分数
     * @param id
     * @param score
     * @return
     */
    Work updateScore(Long id, int score);

    /**
     * 综合查询
     * @param itemId 实验id
     * @param studentName 学生姓名
     * @param studentSno 学生学号
     * @param pageable
     * @return
     */
    Page<Work> getAll(Long itemId, String studentName, String studentSno, Boolean reviewed, @NotNull Pageable pageable);

    /**
     * 判断是否为老师
     * @return 是返回ture， 不是返回false
     */
    boolean isTeacher();
//
//    Work getNextNotReviewedWork();
}

File: workHomeWebApp/src/app/service/work.service.ts
import {Injectable} from '@angular/core';
import {Observable} from 'rxjs';
import {Work} from '../common/work';
import {HttpClient} from '@angular/common/http';
import {Assert} from '../utils';
import {Page} from '../base/page';

@Injectable({
  providedIn: 'root'
})
export class WorkService {
  public static workPage = 0;
  private url = 'work';

  constructor(private httpClient: HttpClient) {
  }

  getAllOfCurrentStudent(): Observable<Array<Work>> {
    return this.httpClient.get<Array<Work>>(`${this.url}/getAllOfCurrentStudent`);
  }

  getAll(params: {page: number, size: number, studentName?: string, studentSno?: string, itemId?: number}):
    Observable<Page<Work>> {
    const _params = {
      page: params.page.toLocaleString(),
      size: params.size.toLocaleString(),
      studentName: params.studentName ? params.studentName : null,
      studentNo: params.studentSno ? params.studentSno : null,
      // reviewed: params.reviewed === undefined ? 'null' : params.reviewed ? '1' : '0',
      itemId: params.itemId ? params.itemId.toLocaleString() : null
    };
    // console.log(_params.reviewed);
    return this.httpClient.get<Page<Work>>(`${this.url}/getAll`, {params: _params});
  }

  /**
   * 根据作业ID获取指定的作业
   * @param id 作业ID
   */
  getById(params: { id: number }): Observable<Work> {
    const _params = {id: params.id.toLocaleString()};
    return this.httpClient.get<Work>(`${this.url}/${params.id}`, {params: _params});
  }

  /**
   * 更新成绩
   * @param id 作业ID
   * @param work Work
   */
  updateScore(params: { id: number, work: Work }): Observable<Work> {
    const _params = params;
    return this.httpClient.put<Work>(`${this.url}/updateScore/${_params.id}`, _params.work);
  }

  /**
   * 更新前台学生作业
   * @param id 作业ID
   * @param work 作业
   */
  updateOfCurrentStudent(id: number, work: Work): Observable<Work> {
    Assert.isNumber(id, '未接收到有效的id值 ');
    Assert.isDefined(work, '未接收到有效的work值');
    return this.httpClient.put<Work>(this.url + '/updateOfCurrentStudent/' + id, work);
  }

  deleteAttachment(workId: number, attachmentId: number): Observable<void> {
    return this.httpClient.delete<void>(this.url + '/deleteAttachment/' + workId + '/' + attachmentId);
  }

  getByItemIdOfCurrentStudent(itemId: number): Observable<Work> {
    const params: { [key: string]: any } = {itemId};
    return this.httpClient.get<Work>(this.url + '/getByItemIdOfCurrentStudent', {params});
  }

  getNextNotReviewedWork(): Observable<Work> {
    return this.httpClient.get<Work>(this.url + '/getNextNotReviewedWork');
  }
}
